beetsplug/bpd/__init__.py

```diff
@@ -708,7 +834,10 @@ def run(self):
                 except BPDClose:
                     # Command indicates that the conn should close.
                     self.sock.close()
+                    self.disconnect()  # Client explicitly closed.
                     return
+                except BPDIdle as e:
+                    yield bluelet.call(self.poll_notifications(e.subsystems))
```

Comment:
Sorry to be slow here, but I don't quite see why the `BPDIdle` exception is necessary. It seems like the idea is that we need to `bluelet.call` the `poll_notifications` method from here, instead of directly calling `poll_notifications` from within the `cmd_idle` handler—but I don't quite follow why *that* is necessary. Can you please explain a little more?

Reply:
I haven't fully wrapped by head around coroutines yet and I quite possibly missed a simpler implementation. I'll have another look because I might be overcomplicating things here.

The `idle` command is fundamentally different to the others though because whereas normally the server waits for a request and then sends a response, in this case it has to delay the response and start waiting on _either_ a new request from the same connection (`noidle`) or an event triggered by a request from a different connection/the player/etc..

The `Server.cmd_*` methods are all called from within `Connection.run` in these lines (where `func` is the bound method like `cmd_idle`):
```python
            args = [conn] + self.args
            results = func(*args)
            if results:
                for data in results:
                    yield conn.send(data)
```
I thought this meant that the command methods can only `yield` bytes to be sent over the connection, but aren't able to hook into the bluelet event loop in any other way or prevent the server from sending an immediate response. The method isn't called with `bluelet.run` so it can at most be a standard generator instead of a coroutine itself right? The new exception seemed necessary to break out of that flow.

Reply:
You're very right! Yes, that answers my question—the `cmd_*` methods are all string generators, not ordinary Bluelet coroutines. Thanks for reminding me of the way my own code works! :smiley: